perftest/PerfTestSocketHandler.cpp

```diff
@@ -0,0 +1,188 @@
+{
+    LOG_DBG("PerfTestSocketHandler sendMessage: " + message);
+    _outgoingMessages.emplace(message);
+}
+
+bool PerfTestSocketHandler::hasIncomingMessage(const std::string &substring)
+{
+    // Checks all messages for substring and clears queue
+    while (!_incomingMessages.empty()) {
+        std::string message = _incomingMessages.front();
+        _incomingMessages.pop();
+        if (message.find(substring) != std::string::npos) {
+            return true;
+        }
+    }
+    return false;
+}
+
+void PerfTestSocketHandler::connect(const std::string &filePath)
+{
+    LOG_DBG("PerfTestSocketHandler connect: " << _server << " file: " << filePath);
+
+    std::string fileUri;
+    std::string fileabs = Poco::Path(filePath).makeAbsolute().toString();
+    Poco::URI::encode("file://" + fileabs, ":/?", fileUri);
+
+    std::string wrap;
+    Poco::URI::encode(fileUri, ":/?", wrap); // double encode.
+    std::string serverUri = _server + "/cool/" + wrap + "/ws";
+
+    LOG_DBG("Connecting to: " + serverUri);
+    std::cerr << "Connecting to: " << serverUri << std::endl;
+    // Cannot implement shared_from_this on PerfTestSocketHandler because it
+    // is already implemented on ProtocolHandlerInterface. So do this cast instead.
+    std::shared_ptr<WebSocketHandler> ptr_to_this =
+        std::dynamic_pointer_cast<WebSocketHandler>(shared_from_this());
+    _poll.insertNewWebSocketSync(Poco::URI(serverUri), ptr_to_this);
+    _poll.startThread();
+    LOG_DBG("PerfTestSocketHandler Connected");
+}
+
+void PerfTestSocketHandler::loadDocument(const std::string &filePath)
+{
+    LOG_DBG("PerfTestSocketHandler loadDocument: " + filePath);
+    std::cerr << "Loading document: " << filePath << std::endl;
+
+    sendMessage("coolclient 0.1 1713197290182 836");
+
+    std::string fileUri;
+    std::string fileabs = Poco::Path(filePath).makeAbsolute().toString();
+    Poco::URI::encode("https://localhost:9980/wopi/files" + fileabs, ":/?", fileUri);
+
+    sendMessage("load url=" + fileUri + " deviceFormFactor=desktop");
+    // clientvisiblearea and clientzoom are necessary for tiles to be rendered and sent
+    sendMessage("clientvisiblearea x=0 y=0 width=99999 height=99999 splitx=0 splity=0");
+    sendMessage("clientzoom tilepixelwidth=256 tilepixelheight=256 tiletwipwidth=3840 tiletwipheight=3840");
+}
+
+void PerfTestSocketHandler::shutdown()
+{
+    LOG_DBG("PerfTestSocketHandler shutdown");
+    _shutdown = true;
+    _poll.joinThread();
+    LOG_DBG("PerfTestSocketHandler done");
+}
+
+int PerfTestSocketHandler::getPollEvents(std::chrono::steady_clock::time_point now, int64_t &timeoutMaxMicroS)
+{
+    LOG_TRC("PerfTestSocketHandler getPollEvents");
+    if (_connecting) {
+        LOG_TRC("PerfTestSocketHandler getPollEvents (Waiting for outbound connection)");
+        return POLLOUT;
+    }
+
+    while (!_outgoingMessages.empty()) {
+        std::string message = _outgoingMessages.front();
+        _outgoingMessages.pop();
```

Comment:
How do we avoid pushing all the session's events down the socket in one whack ? That is not going to simulate real use; it seems you removed the timestamp handling from the trace replay code here - why ?